### PR TITLE
Data scripts: now includes the hash

### PR DESCRIPTION
**Pull request by michaelpj**
The `Sealed` argument should include the hash, now it does. Users can
cross-reference this against the `PendingTx` to
work out which output it comes from. Example in `StateMachine`.

Note that this is probably unnecessary if you only care that it is *a*
data script, but often you want to check that it's on the right output.

I also had to do a bit of shuffling things around modules, mostly just moving the definition of the various `*Hash` types to `Scripts` with some cascading changes.

### CODE REVIEW

**Comment by michaelpj** [comment]
Also, the changes I made to the state machine validator make it significantly more secure - it also previously didn't check that there was precisely one output that continues the machine, which can lead to weirdness.

File: plutus-wallet-api/ledger/Ledger/Validation.hs [hunk 1]
```diff
@@ -135,6 +122,25 @@ data PendingTx = PendingTx
     -- ^ Hash of the pending transaction (excluding witnesses)
     } deriving (Generic)
 
+{-# INLINABLE findDataScriptOutputs #-}
+-- | Look up a 'DataScriptHash' in the transaction outputs, returning the indexs of the outputs that have
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Validation.hs]:
```suggestion
-- | Look up a 'DataScriptHash' in the transaction outputs, returning the indices of the outputs that have
```

File: plutus-wallet-api/src/Language/PlutusTx/StateMachine.hs [hunk 2]
```diff
@@ -38,10 +40,23 @@ mkRedeemer i ss = (i, ss)
 {-# INLINABLE mkValidator #-}
 -- | Turn a transition function 's -> i -> s' into a validator script.
 mkValidator :: Eq s => StateMachine s i -> StateMachineValidator s i
-mkValidator (StateMachine trans) currentState (input, unseal -> newState) _ =
+mkValidator (StateMachine trans) currentState (input, unseal -> HashedDataScript newState hsh) ptx =
     let
+        vsOutput = uniqueElement (findContinuingOutputs ptx)
+        dsOutputs = findDataScriptOutputs hsh ptx
+        dataScriptOk = case vsOutput of
+            -- It is *not* okay to have multiple outputs with the current validator script, that allows "spliting" the machine.
```
Comment (j-mueller) [inline plutus-wallet-api/src/Language/PlutusTx/StateMachine.hs]:
```suggestion
            -- It is *not* okay to have multiple outputs with the current validator script, that allows "splitting" the machine.
```

**Comment by michaelpj** [comment]
Sigh, another mystery non-determinism in the output.

**Comment by michaelpj** [comment]
I don't have the spirit to debug this today, I'll come back to it on Monday.

**Comment by michaelpj** [comment]
Darwin builds are blocked, sadly. This now works, although I have this weird performance issue.

**Comment by michaelpj** [comment]
Mystery slowdown went away, I think it was just my machine being overloaded, hooray. It's also no slower on Hydra.

**Comment by michaelpj** [comment]
The second commit fixes a source of instability (not non-determinism, fortunately), that gave us different results for things in Nix and in Cabal (sensitivity to the exact hash of a module). So that seems to be working now, but it gives rather large diffs since everything gets reordered.

**Comment by michaelpj** [comment]
Okay, this should work now!

**Comment by michaelpj** [comment]
Here goes.

**Pull request merged**